packages/core-contracts/contracts/proxy/UUPSImplementation.sol

```diff
@@ -4,19 +4,28 @@ pragma solidity ^0.8.0;
 import "../utils/ContractUtil.sol";
 import "../utils/TransformUtil.sol";
 import "../common/CommonErrors.sol";
+import "./ProxyStorage.sol";
 
-abstract contract UniversalProxyImplementation is ContractUtil, TransformUtil, CommonErrors {
+abstract contract UUPSImplementation is ProxyStorage, ContractUtil, TransformUtil, CommonErrors {
     error ImplementationIsSterile(address implementation);
 
     event Upgraded(address implementation);
 
-    function _setImplementation(address newImplementation) internal virtual;
+    function _setImplementation(address newImplementation) internal virtual {
+        _proxyStorage().implementation = newImplementation;
+    }
 
-    function _getImplementation() internal view virtual returns (address);
+    function _getImplementation() internal view virtual returns (address) {
+        return _proxyStorage().implementation;
+    }
 
-    function _setSimulatingUpgrade(bool simulatingUpgrade) internal virtual;
+    function _setSimulatingUpgrade(bool simulatingUpgrade) internal virtual {
+        _proxyStorage().simulatingUpgrade = simulatingUpgrade;
+    }
 
-    function _getSimulatingUpgrade() internal view virtual returns (bool);
+    function _getSimulatingUpgrade() internal view virtual returns (bool) {
+        return _proxyStorage().simulatingUpgrade;
+    }
 
     function upgradeTo(address newImplementation) public virtual;
```

Comment:
upgradeTo is still virtual to allow the implementation to use any modifier to protect it

Reply:
Yeah, I saw this earlier today. I didn't quite get the reason for having `upgradeTo` and `_upgradeTo`. As far as I understand modifiers don't affect inheritance so you don't need this. Or is there some other reason?